### PR TITLE
API v2: address nft

### LINKED ISSUE #8231: Token ids for ERC-721
https://github.com/blockscout/blockscout/issues/3050
https://github.com/blockscout/blockscout/issues/6068
https://github.com/blockscout/blockscout/issues/1074

### PR DESCRIPTION
**Pull request by nikitosing**
Close #8231 

## Changelog
- Add filter to `/api/v2/tokens/{address_hash}/instances?holder_address_hash={holder_address}`
- Add `/api/v2/addresses/{address_hash}/nft` and `/api/v2/{address_hash}/nft/collections`

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/address/current_token_balance.ex [hunk 1]
```diff
@@ -49,6 +49,9 @@ defmodule Explorer.Chain.Address.CurrentTokenBalance do
     field(:token_id, :decimal)
     field(:token_type, :string)
     field(:fiat_value, :decimal, virtual: true)
+    field(:distinct_token_instances_count, :integer, virtual: true)
+    field(:token_ids, {:array, :decimal}, virtual: true)
+    field(:preloaded_token_instances, {:array, :any}, virtual: true)
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/address/current_token_balance.ex:54]: Please, upate the type spec

File: apps/explorer/lib/explorer/chain/token/instance.ex [hunk 2]
```diff
@@ -33,6 +35,7 @@ defmodule Explorer.Chain.Token.Instance do
     field(:error, :string)
     field(:owner_updated_at_block, :integer)
     field(:owner_updated_at_log_index, :integer)
+    field(:current_token_balance, :any, virtual: true)
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/token/instance.ex:38]: Please, update the type spec

**Review by sl1depengwyn — CHANGES_REQUESTED** [review]

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_controller.ex [hunk 3]
```diff
@@ -87,6 +88,42 @@ defmodule BlockScoutWeb.API.V2.TokenController do
     end
   end
 
+  def instances(
+        conn,
+        %{"address_hash_param" => address_hash_string, "holder_address_hash" => holder_address_hash} = params
+      ) do
+    with {:format, {:ok, address_hash}} <- {:format, Chain.string_to_address_hash(address_hash_string)},
+         {:ok, false} <- AccessHelper.restricted_access?(address_hash_string, params),
+         {:not_found, {:ok, token}} <- {:not_found, Chain.token_from_address_hash(address_hash, @api_true)},
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_controller.ex:97]: I think it's better to factor out such refactoring to a separate PR, in order to not to overload PR

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_controller.ex [hunk 4]
```diff
@@ -87,6 +88,42 @@ defmodule BlockScoutWeb.API.V2.TokenController do
...
+         {:not_found, false} <- {:not_found, Chain.is_erc_20_token?(token)},
+         {:format, {:ok, holder_address_hash}} <- {:format, Chain.string_to_address_hash(holder_address_hash)} do
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_controller.ex]: Nice catch! Fixed

**Review by sl1depengwyn — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

**Pull request merged**